### DISCUSSION TITLE
Prediction using Tensors to avoid detachment

### DISCUSSION THREAD

**Discussion by damandak** [Nixtla/neuralforecast#1013] [Q&A]
Hi,

I'm working on a explainable AI project using N-BEATSx and N-HiTS models with exogenous variables, using gradient-based perturbations. For this, I need the predict functions to receive pytorch tensors, but the predict functions of N-BEATSx and N-HiTS models accept Pandas DataFrames as input. This breaks the computational graph when converting from tensors. This loss of gradient information makes gradient-based perturbations impossible.

I need a way to run predictions directly on PyTorch tensors to maintain the computational graph and enable gradient computation. Is there an existing method or a recommended approach to achieve this within the library?

Any guidance or suggestions would be greatly appreciated.

Thank you!

**Comment by jmoralez** [comment]
Hey. The dataframes interface is a convenience, if you want to you can directly use the forward methods ([NBEATSx](https://github.com/Nixtla/neuralforecast/blob/1e0dffc1bfb72140f693eb66e5c7af77d4dd646e/neuralforecast/models/nbeatsx.py#L505), [NHITS](https://github.com/Nixtla/neuralforecast/blob/1e0dffc1bfb72140f693eb66e5c7af77d4dd646e/neuralforecast/models/nhits.py#L389))  which take dicts of tensors and return a tensor.

If you need utilities for going from a dataframe to batches you can use [TimeSeriesDataset.from_df](https://github.com/Nixtla/neuralforecast/blob/1e0dffc1bfb72140f693eb66e5c7af77d4dd646e/neuralforecast/tsdataset.py#L280) and the [data module](https://github.com/Nixtla/neuralforecast/blob/1e0dffc1bfb72140f693eb66e5c7af77d4dd646e/neuralforecast/tsdataset.py#L375).

**Reply by damandak** [reply]
Thank you! Using the forward methods, those work great!
